### PR TITLE
feat: add logging hook, rm logging from evaluation

### PR DESCRIPTION
**Pull request by liran2000**
Implements:  [[FEATURE] Implement Logging Hook](https://github.com/open-feature/go-sdk/issues/284#top)

This PR removes logging from the main evaluation code path, with a few exceptions, like unhandled errors in event handlers. These cases are printed with default _slog_.
It also adds a LoggingHook as per the spec link above, using [slog](https://pkg.go.dev/log/slog).

_logr_ usage is removed, but exported APIs remains for backward compatibility.

[Java implementation](https://github.com/open-feature/java-sdk/pull/1084)

Closes: https://github.com/open-feature/go-sdk/issues/284

@toddbaert @Kavindu-Dodan

### CODE REVIEW

**Comment by toddbaert** [comment]
@liran2000 sorry I missed this! I will review today. Looks like there's some CI failures though.

**Comment by liran2000** [comment]
> @liran2000 sorry I missed this! I will review today. Looks like there's some CI failures though.

Thanks, this PR requires GO version to be upgraded to 1.21 before proceeding, hence the failures.

**Comment by toddbaert** [comment]
I've opened: https://github.com/open-feature/go-sdk/pull/294

File: openfeature/client.go [hunk 1]
```diff
@@ -52,25 +51,24 @@ var _ IClient = (*Client)(nil)
...
 		api:               apiRef,
 		clientEventing:    eventRef,
-		logger:            log,
 		metadata:          ClientMetadata{name: name},
 		hooks:             []Hook{},
 		evaluationContext: EvaluationContext{},
 	}
 }
 
+// Deprecated
```
Comment (toddbaert) [inline openfeature/client.go:67]: :+1:

**Review by toddbaert — APPROVED** [review]

File: openfeature/hooks/logging_hook.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	VARIANT_KEY            = "variant"
+	VALUE_KEY              = "value"
+)
+
+type LoggingHook struct {
+	includeEvaluationContext bool
+	logger                   *slog.Logger
+}
+
+func NewLoggingHook(logger *slog.Logger, includeEvaluationContext bool) (*LoggingHook, error) {
```
[the remark this answers is not in the document]
Comment (liran2000) [reply openfeature/hooks/logging_hook.go:28]: _slog_ logger can be used as global, or not. From the [docs](https://pkg.go.dev/log/slog):

> For more control over the output format, create a logger with a different handler.

That gives the user more control without limiting it to use only the global one.

[the remark this answers is not in the document]
Comment (toddbaert) [reply openfeature/hooks/logging_hook.go:28]: Ya, I was just saying that we can also have a constructor that doesn't require the instance, but it's OK not to; it's easy to get one.

[the remark this answers is not in the document]
Comment (liran2000) [reply openfeature/hooks/logging_hook.go:28]: checking, and also the CI

**Comment by toddbaert** [comment]
@liran2000 I've merged the Go 1.2.1 change, there's a few lint/CI failures.

File: openfeature/hooks/logging_hook.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+type LoggingHook struct {
+	includeEvaluationContext bool
+	logger                   *slog.Logger
+}
+
+func NewLoggingHook(includeEvaluationContext bool) (*LoggingHook, error) {
+	return NewCustomLoggingHook(slog.Default(), includeEvaluationContext)
+}
+
+func NewCustomLoggingHook(logger *slog.Logger, includeEvaluationContext bool) (*LoggingHook, error) {
```
Comment (toddbaert) [inline openfeature/hooks/logging_hook.go:32]: One last nit: can you reverse the param order so `includeEvaluationContext bool` is always first? :sweat_smile:

Comment (liran2000) [reply openfeature/hooks/logging_hook.go]: done

File: README.md [hunk 4]
```diff
@@ -175,21 +175,14 @@ value, err := client.BooleanValue(
...
-var l logr.Logger
-l = integratedlogr.New() // replace with your chosen integrator
-
-openfeature.SetLogger(l) // set the logger at global level
+#### Logging Hook
 
-c := openfeature.NewClient("log").WithLogger(l) // set the logger at client level
-```
+The GO SDK includes a `LoggingHook`, which logs detailed information at key points during flag evaluation, using [slog](https://pkg.go.dev/log/slog) structured logging API.
+This hook can be particularly helpful for troubleshooting and debugging; simply attach it at the global, client or invocation level and ensure your log level is set to "debug".
```
[the remark this answers is not in the document]
Comment (liran2000) [reply README.md:184]: added

**Comment by toddbaert** [comment]
Thanks again @liran2000

**Pull request merged**